N°4947 Fix \EMail::LoadConfig

piRGoif opened the pull request:
In `\EMail::SendSynchronous` we are calling `\EMail::LoadConfig`.
This method uses the `ITOP_DEFAULT_CONFIG_FILE` constant, and as a consequence is always loading config using the production environment.

This PR changes the loading and now uses the standard `\utils::GetConfig` method !

Config was saved in the `\EMail::$m_oConfig` attribute : it is only written there.
The `ITOP_DEFAULT_CONFIG_FILE` constant is only used in the toolkit and `sources/application/status/status.inc.php`.  All of those usages seems legit to me so I'm not changing them in this PR.

Thanks to @jbostoen for the bug report (see https://github.com/Combodo/iTop/pull/223#issuecomment-1064854430) !

piRGoif committed:
Fix \EMail::LoadConfig

Hipska commented on core/email.class.inc.php:
```diff
@@ -41,11 +41,10 @@ class EMail
-		{
-			self::$m_oConfig = new Config($sConfigFile);
+		if (is_null(self::$m_oConfig)) {
+			self::$m_oConfig = utils::GetConfig();
```
Why using `utils::GetConfig()` while the constructor uses `MetaModel::GetConfig()`? This seems inconsistent..

piRGoif replied:
Ouch, didn't see that, thanks !

piRGoif replied:
Fixed in 12951f8a

piRGoif committed:
Fix config use in \EMail::__construct

Hipska commented on core/email.class.inc.php:
```diff
@@ -55,7 +54,13 @@ public function __construct()
-		$this->SetRecipientFrom(MetaModel::GetConfig()->Get('email_default_sender_address'), MetaModel::GetConfig()->Get('email_default_sender_label'));
+
+		$this->LoadConfig();
+		$oConfig = self::$m_oConfig;
+		$this->SetRecipientFrom(
+			$oConfig->Get('email_default_sender_address'),
+			$oConfig->Get('email_default_sender_label')
+		);
```
```suggestion
		$this->SetRecipientFrom(
			static::$m_oConfig->Get('email_default_sender_address'),
			static::$m_oConfig->Get('email_default_sender_label')
		);
```

piRGoif replied:
I found my solution easier to read...
Guess we'll see what other devs will say during the review

Hipska replied:
What about removing the `LoadConfig` and `$m_oConfig` completely and do this in construct then: `$oConfig = utils::GetConfig()`. It doesn't make much sense to have it as a static var.

piRGoif replied:
In the legacy code, there were lots of static attributes to save the config object (it was the case in the utils class before 2.7.0 and that was the main reason why I was calling it a mess :o) ). I guess this was done for performance reasons ? 
Removing `\EMail::$m_oConfig` is something we'll have to discuss during this PR dev review indeed !
I have some doubt personally, as the class can be called multiple times without the MetaModel loaded, so this would mean opening the config file each time...

There are no reasons to keep the `\EMail::LoadConfig` call in `\EMail::SendSynchronous` though, so I'm removing it ! (commit 1ab3a51f)

Hipska replied:
It indeed makes sense to not to have to load the config multiple times if metamodel isn't loaded yet. I agree to keep it static for now.

PS; just make sure it's using `static::$m_oConfig` instead of `self::$m_oConfig` everywhere.

Hipska replied:
Sorry to come back on this, but what do you think of the following?
```php
$oConfig = $this->LoadConfig();
```

piRGoif replied:
Done in 99ab128d :)
Closing this conversation

Hipska commented on core/email.class.inc.php:
```diff
@@ -41,11 +41,10 @@ class EMail
-		if (is_null(self::$m_oConfig))
-		{
-			self::$m_oConfig = new Config($sConfigFile);
+		if (is_null(self::$m_oConfig)) {
+			self::$m_oConfig = utils::GetConfig();
```
```suggestion
		if (is_null(static::$m_oConfig)) {
			static::$m_oConfig = utils::GetConfig();
```

Hipska reviewed (changes requested):
When the LoadConfig is already done in the constructor, there is no need to do it on other places anymore..

piRGoif committed:
Replace self by static

Co-authored-by: Thomas Casteleyn <thomas.casteleyn@super-visions.com>

piRGoif committed:
Removing LoadCOnfig call in \EMail::SendSynchronous
This isn't a static method, and this call is already done in the class constructor

piRGoif committed:
Change uses of m_oConfig attribute from self to static

piRGoif commented:
Indeed, there were quite a few remaining uses of the attribute with the "self" keyword. I've changed them in 8640918d
Thanks !

Molkobain commented on core/email.class.inc.php:
```diff
@@ -41,11 +41,10 @@ class EMail
 	protected $m_aData; // For storing data to serialize
 
-	public function LoadConfig($sConfigFile = ITOP_DEFAULT_CONFIG_FILE)
+	public function LoadConfig()
```
Can you just add an `@since` to notify of the param removal?

piRGoif replied:
Waiting for the functional review so that we know in which version this will be merged

piRGoif replied:
Done in 9ea94415

Molkobain commented:
Accepted during technical review.

Molkobain commented:
Accepted in functional review.
Will be merged on support/2.7 branch **only once the 2.7.7 is out**, so we can test it on 3.0.2 first.

piRGoif committed:
Added `@since`
Now we know which are the targeted versions

piRGoif committed:
LoadConfig now returns the m_oConfig value
Thanks Hipska for the suggestion !

piRGoif commented:
The Email class was totally refactored in iTop 2.7.7 : 

* eb1d56f43942b3f49275005073919a701f13d109
* 754946bf62fcc9f62ddaf8120951b6c6127001e1
* 622f40c06cf9d3dbe170416498c4906dd1314318

Rebasing this branch cannot be done anymore. I'm closing this PR and opening a new one.

Molkobain commented:
Can we delete the branch?

piRGoif commented:
Now that I finished the new PR, yes :)
